File: Topics/Queue&Deque/933.py
from collections import deque

class RecentCounter:

    def __init__(self):
        self.counter = deque()
        self.cnt = 0

    def ping(self, t: int) -> int:
        self.counter.append(t)
        self.cnt += 1
        while self.counter[-1] - 300 > self.counter[0]:
             self.cnt -= 1
             self.counter.popleft()
        return self.cnt


# Your RecentCounter object will be instantiated and called as such:
# obj = RecentCounter()
# param_1 = obj.ping(t)

File: Topics/Queue&Deque/933.ts
class RecentCounter {
    counter: number;
    queue: number[];
    p1: number;
    p2: number;
    constructor() {
        this.counter = 0
        this.queue = []
        this.p1 = 0
        this.p2 = 0
    }

    ping(t: number): number {
        while(this.queue.length > this.p1 && t - this.queue[this.p1] > 3000) {
            this.p1 += 1
        }
        this.queue.push(t)
        this.p2 += 1
        return this.p2 - this.p1
    }
}



var obj = new RecentCounter()
console.log(obj.ping(1))
console.log(obj.ping(2))
console.log(obj.ping(3000))
console.log(obj.ping(3002))
